File: 0399-evaluate-division/0399-evaluate-division.php
class Solution {

    private $graph = [];

    /**
     * @param String[][] $equations
     * @param Float[] $values
     * @param String[][] $queries
     * @return Float[]
     */
    function calcEquation($equations, $values, $queries) {
        for ($i = 0; $i < count($equations); $i++) {
            $equation = $equations[$i];
            if (!isset($this->graph[$equation[0]])) {
                $this->graph[$equation[0]] = [];
            }
            if (!isset($this->graph[$equation[1]])) {
                $this->graph[$equation[1]] = [];
            }
            $this->graph[$equation[0]][$equation[1]] = $values[$i];
            $this->graph[$equation[1]][$equation[0]] = 1 / $values[$i];
        }

        $result = [];
        for ($i = 0; $i < count($queries); $i++) {
            $query = $queries[$i];
            array_push($result, $this->DFS($query[0], $query[1], []));
        }
        return $result;
    }

    private function DFS($start, $end, $visited) {
        if (!isset($this->graph[$start])) return -1.0;
        if (isset($this->graph[$start][$end])) return $this->graph[$start][$end];

        $visited[$start] = true;
        foreach ($this->graph[$start] as $neighbor => $value) {
            if (isset($visited[$neighbor])) continue;
            $product = $this->DFS($neighbor, $end, $visited);
            if ($product != -1.0) return $value * $product;
        }
        return -1.0;
    }
}

File: 0399-evaluate-division/0399-evaluate-division.cs
public class Solution {
    private Dictionary<string, Dictionary<string, double>> graph = new Dictionary<string, Dictionary<string, double>>();
    
    public double[] CalcEquation(IList<IList<string>> equations, double[] values, IList<IList<string>> queries) {
        for (int i = 0; i < equations.Count; i++) {
            var equation = equations[i];
            if (!graph.ContainsKey(equation[0])) {
                graph[equation[0]] = new Dictionary<string, double>();
            }
            if (!graph.ContainsKey(equation[1])) {
                graph[equation[1]] = new Dictionary<string, double>();
            }
            graph[equation[0]][equation[1]] = values[i];
            graph[equation[1]][equation[0]] = 1 / values[i];
        }
        
        double[] result = new double[queries.Count];
        for (int i = 0; i < queries.Count; i++) {
            var query = queries[i];
            result[i] = DFS(query[0], query[1], new HashSet<string>());
        }
        return result;
    }
    
    private double DFS(string start, string end, HashSet<string> visited) {
        if (!graph.ContainsKey(start)) return -1.0;
        if (graph[start].ContainsKey(end)) return graph[start][end];
        
        visited.Add(start);
        foreach (var neighbor in graph[start].Keys) {
            if (!visited.Contains(neighbor)) {
                double productWeight = DFS(neighbor, end, visited);
                if (productWeight != -1) return graph[start][neighbor] * productWeight;
            }
        }
        return -1.0;
    }
}
